routing/matcher.go

```diff
@@ -283,15 +290,85 @@ func moveToSubtreeIfExists(subtree *pathMatcher, paths map[string]*pathMatcher,
 	}
 
 	subtree.leaves = append(subtree.leaves, pm.leaves...)
-
 	delete(paths, path)
 }
 
+// when a path subtree and a path with free wildcard has the same root,
+// we need to merge them to the same leaf, but we need to preserve the
+// original parameter name. The tree accepts only a single wildcard
+// param name for a given position, so we need to store the original
+// param name elsewhere (subtreeMergeControl).
+func mergeWithSubtreeIfConflicting(subtrees, paths map[string]*pathMatcher) {
+	subtreesUnnamed := make(map[string]*pathMatcher)
+	subtreesFullPaths := make(map[string]string)
+	for stp, stm := range subtrees {
+		// trim the param name, if exists
+		var subtreePath string
+		stpParam := freeWildcardRx.FindString(stp)
+		if stpParam == "" {
+			subtreePath = stp
+		} else {
+			subtreePath = stp[:len(stp)-len(stpParam)]
+		}
+
+		if subtreePath == "" {
+			subtreePath = "/"
+		}
+
+		subtreesUnnamed[subtreePath] = stm
+		subtreesFullPaths[subtreePath] = stp
+	}
+
+	var removeFromPaths []string
+	for p, pm := range paths {
+		fwm := freeWildcardRx.FindString(p)
+		if fwm == "" {
+			continue
+		}
+
+		subRoot := p[:len(p)-len(fwm)]
+		if subRoot == "" {
+			subRoot = "/"
+		}
+
+		stm, ok := subtreesUnnamed[subRoot]
+		if !ok {
+			continue
+		}
+
+		fromParam := stm.freeWildcardParam
+		if fromParam == "" {
+			fromParam = "*"
+		}
+
+		for _, l := range pm.leaves {
+			l.subtreeMergeControl.ignoreSubtreeRoot = true
+			l.subtreeMergeControl.subtreeRoot = subRoot
+			l.subtreeMergeControl.mapParamNameFrom = fromParam
+			l.subtreeMergeControl.mapParamNameTo = pm.freeWildcardParam
+		}
+
+		pm.leaves = append(pm.leaves, stm.leaves...)
+		pm.freeWildcardParam = stm.freeWildcardParam
+		subtrees[subtreesFullPaths[subRoot]] = pm
+		removeFromPaths = append(removeFromPaths, p)
+	}
+
+	for _, p := range removeFromPaths {
+		delete(paths, p)
+	}
+}
+
 func moveConflictingToSubtree(subtrees, paths map[string]*pathMatcher) {
 	for p, stm := range subtrees {
 		moveToSubtreeIfExists(stm, paths, p)
+
+		// TODO: this may mean that a non-subtree route will match even when ignore-trailing-match is
```

Comment:
non-subtree route is a route having `Path` and not `PathSubtree` ?
Do you want to fix this in this PR?